qp_shotgun/kneaddata/kneaddata.py

```diff
@@ -0,0 +1,278 @@
+                            rev_fp))
+
+        used_prefixes.add(run_prefix)
+
+    return(samples)
+
+
+def _format_kneaddata_params(parameters):
+    params = []
+
+    for param in sorted(parameters):
+        value = parameters[param]
+
+        if value is True:
+            params.append('--%s' % param)
+        elif value is False:
+            continue
+        elif value and value != 'default':
+            params.append('--%s %s' % (param, value))
+
+    param_string = ' '.join(params)
+
+    return(param_string)
+
+
+def generate_kneaddata_commands(forward_seqs, reverse_seqs, map_file,
+                                out_dir, parameters):
+    """Generates the KneadData commands
+
+    Parameters
+    ----------
+    forward_seqs : list of str
+        The list of forward seqs filepaths
+    reverse_seqs : list of str
+        The list of reverse seqs filepaths
+    map_file : str
+        The path to the mapping file
+    out_dir : str
+        The job output directory
+    parameters : dict
+        The command's parameters, keyed by parameter name
+
+    Returns
+    -------
+    cmds: list of str
+        The KneadData commands
+    samples: list of tup
+        list of 4-tuples with run prefix, sample name, fwd read fp, rev read fp
+
+    Notes
+    -----
+    Currently this is requiring matched pairs in the make_read_pairs_per_sample
+    step but implicitly allowing empty reverse reads in the actual command
+    generation. This behavior may allow support of situations with empty
+    reverse reads in some samples, for example after trimming and QC.
+    """
+    # we match filenames, samples, and run prefixes
+    samples = make_read_pairs_per_sample(forward_seqs, reverse_seqs, map_file)
+
+    cmds = []
+
+    param_string = _format_kneaddata_params(parameters)
+    for run_prefix, sample, f_fp, r_fp in samples:
+        r_fp_str = ' --input "%s"' % r_fp if r_fp is not None else ""
+        cmds.append('kneaddata --input "%s"%s --output "%s" '
+                    '--output-prefix "%s" %s' % (
+                        f_fp, r_fp_str, join(out_dir, run_prefix), run_prefix,
+                        param_string))
+
+    return cmds, samples
+
+
+def _run_commands(qclient, job_id, commands, msg):
+    for i, cmd in enumerate(commands):
+        qclient.update_job_step(job_id, msg % i)
+        std_out, std_err, return_value = system_call(cmd)
+        if return_value != 0:
+            error_msg = ("Error running KneadData:\nStd out: %s\nStd err: %s"
+                         "\n\nCommand run was:\n%s"
+                         % (std_out, std_err, cmd))
+            return False, error_msg
+
+    return True, ""
+
+
+def _per_sample_ainfo(out_dir, samples):
+    pf = []
+    r1f = []
+    r2f = []
+    for rp, _, _, _ in samples:
+        smd = partial(join, out_dir, rp)
+
+        fname = smd('%s.trimmed.1.fastq' % rp)
+        if not exists(fname):
+            raise ValueError("%s doesn't exist" % fname)
```

Comment:
Ehhhh not here because I think this should be getting raw sequence, so definitely everything should be paired. But e.g. for Humann2 processing or the like.

Probably stuff that will be smoothed over by a more useful fastq representation / file format / object...